libnavigation-core/src/main/java/com/mapbox/navigation/core/sensor/SensorData.kt

```diff
@@ -0,0 +1,176 @@
+                    com.mapbox.navigator.Weather.WET_ROAD -> WET_ROAD
+                }
+            }
+        }
+    }
+
+    /**
+     * Lane information obtained from sensors
+     *
+     * @param currentLaneIndex the index of the current lane. For right-hand traffic 1 is the very right lane and so on.
+     * Should be zero if lane information is not available
+     *
+     * @param laneCount the number of lanes
+     */
+    class Lane(val currentLaneIndex: Int, val laneCount: Int) : SensorData() {
+
+        /**
+         * Indicates whether some other object is "equal to" this one.
+         */
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (javaClass != other?.javaClass) return false
+
+            other as Lane
+
+            if (currentLaneIndex != other.currentLaneIndex) return false
+            return laneCount == other.laneCount
+        }
+
+        /**
+         * Returns a hash code value for the object.
+         */
+        override fun hashCode(): Int {
+            var result = currentLaneIndex
+            result = 31 * result + laneCount
+            return result
+        }
+
+        /**
+         * Returns a string representation of the object.
+         */
+        override fun toString(): String {
+            return "Lane(currentLaneIndex=$currentLaneIndex, laneCount=$laneCount)"
+        }
+    }
+
+    @JvmSynthetic
+    internal fun toNativeSensorData(): com.mapbox.navigator.SensorData {
+        /**
+         * NN requires [com.mapbox.navigator.SensorData.monotonicTimestampNanoseconds] to be based
+         * on the same source as [com.mapbox.navigator.FixLocation.monotonicTimestampNanoseconds],
+         * which is in turn based on [Location#getElapsedRealtimeNanos()](https://developer.android.com/reference/android/location/Location#getElapsedRealtimeNanos()),
+         *
+         * @see [com.mapbox.navigation.core.navigator.toFixLocation]
+         */
+        val elapsedRealtimeNanos = SystemClock.elapsedRealtimeNanos()
+
+        val (type, value) = when (this) {
+            is Weather -> {
+                SensorType.WEATHER to toValue()
+            }
+            is Lane -> {
+                SensorType.LANE to toValue()
+            }
+            else -> error("Unsupported type: $javaClass")
+        }
+        return com.mapbox.navigator.SensorData(type, elapsedRealtimeNanos, value)
+    }
+
+    private fun Weather.toValue(): Value {
+        val order: Long = when (conditionType) {
+            Weather.ConditionType.RAIN -> 0
+            Weather.ConditionType.SNOW -> 1
+            Weather.ConditionType.FOG -> 2
+            Weather.ConditionType.WET_ROAD -> 3
+            else -> error("Unsupported weather condition type: $conditionType")
+        }
+        return Value.valueOf(order)
+    }
+
+    private fun Lane.toValue(): Value {
+        val str = "{\"currentLaneIndex\": $currentLaneIndex, \"laneCount\": $laneCount}"
```

Comment:
Is this expected that lane value is a json? Do we have some agreement on this? Value can also contain map.

Reply:
Great point, thanks for noticing!

@0xDAD could you please take a look? Any guarantee that this format won't be changed?

Reply:
currently we expect json like that
```
{"currentLaneIndex": 1, "laneCount": 3}
```

docs say:

  Lane sensor:
  * value is an object with the following fields: `{"currentLaneIndex": 1, "laneCount": 3}`
  * For right-hand traffic 1 is the very right lane and so on
  * currentLaneIndex == 0 means that lane information is not available

It may change or it may become deprecated. Depends on how NIO uses it. I'd leave that in the form of arbitrary json object just to make it as flexible as possible but it's up to you do decide

Reply:
@0xDAD On platform side we accept object, not json and we convert object to json when ween to pass it from platform to C++ side.

There might be issues if you change expected json format and don't notify us, so this solution doesn't look like reliable. Makes sense to consider changing it on NN side.